pom.xml

```diff
@@ -322,11 +325,41 @@
                 <arg>${cmake.generated.directory}</arg>
               </compilerArgs>
             </configuration>
+            <executions>
+              <!-- compile everything to ensure module-info contains right entries -->
+              <execution>
+                <id>default-compile</id>
+                <configuration>
+                  <release>9</release>
+                  <compileSourceRoots>
+                    <sourceRoot>src/main/java</sourceRoot>
+                    <sourceRoot>src/main/module-info</sourceRoot>
+                  </compileSourceRoots>
+                </configuration>
+              </execution>
+              <!-- recompile everything for target VM except the module-info.java -->
+              <execution>
+                <id>minimal-target-jre-recompile</id>
+                <goals><goal>compile</goal></goals>
+              </execution>
+            </executions>
           </plugin>
         </plugins>
       </build>
     </profile>
 
+    <!-- Convenience profile for users of JDKs past 11 who may want to
+         build the library ignoring no longer supported target version 6 -->
```

Comment:
Does maven offer a way to do this automatically and issue a warning?  If the build fails for JDK12+, we should probably issue a warning and just fix the build system, no?

Reply:
Depends on what are we aiming at I guess.   

As it is now, (before this patch) the project is not buildable on jdk12+, because jdk12+ does not support compilation targeting java version 6 - the error message is quite clear if you attempt such a build. With this patch, users compiling with newer jdk will be automatically accomodated, with the caveat that the resulting artifact is not evenly compatible. I'd assume that users who compile using jdk12+ do not care at all about targeting jre6 and there is nothing about the project that would require enforcing this compatibility.   

Importatnly, users who would build the project themselves are NOT using it as a dependency, and thus presumably modified it somehow anyway. We may as well let them figure out the compilation problem with jdk12+ on their own, but including the profile that does this automatically is just for convienience and future proofing.

We may very well prohibit building with any other version than a cannonical jdk we choose to aim for 100% reproducible builds. This goes back to my original question in #79 - is this library supposed to be BUILD using any jdk, and we are ok with differing results, or should we stick to a single build configuration.

Reply:
> We may as well let them figure out the compilation problem with jdk12+ on their own

I generally come from a different mindset.  Unless there are code compatibilities (such as in the even of Python 2 vs. 3), I'll never fail a build that would otherwise pass.

For example, on macOS running the M1 processor, only proprietary Java flavors are available older than JDK16, so we'd be knowingly failing all builds without special flags.

The same may occur when someone snags the "Latest" JDK from AdoptOpenJDK's page.  If Maven allows these builds to conditionally succeed, that is the preferred approach.

> users who would build the project themselves are NOT using it as a dependency

Chances are, their end goal is.  I'd rather not make any assumptions about intent.

> We may very well prohibit building with any other version than a cannonical jdk we choose to aim for 100% reproducible builds. 

This can certainly be a project goal, but I would never be in favor of passing this burden onto the user.

Reply:
Then we seem to be on the same page. 
After this merge, executing `mvn clean install` without any explicit profile selection on the part of the user will result in the following jar:

```
| Build with JDK | Runs on JRE | 
| 6              | -           | Failed. Maven requires to build with JDK7+
| 7              | 6+          | not compatible with Java Platform Module System
| 8              | 6+          | not compatible with Java Platform Module System 
| 9              | 6+          | Compatible with Java Platform Module System
| 11             | 6+          | Compatible with Java Platform Module System - This should be published to Maven Central
| 12             | 7+          | Compatible with Java Platform Module System
| 15             | 7+          | Compatible with Java Platform Module System

```
I guess we can hack it a bit more to make sure there is a warning message when building with jdk12+, but knowing maven, it takes something like configuring yet another plugin execution, because simple warning is just too simple for xml to handle :) .